### PR TITLE
Review PR for issue #4756

### PR DESCRIPTION
**Pull request by pkarunyu**
Requesting for review of changes suggested in issue #4756

### CODE REVIEW

File: system/database/drivers/mysqli/mysqli_driver.php [hunk 1]
```diff
@@ -484,6 +499,91 @@ public function field_data($table)
...
+	 */
+	public function list_indexes($table)
+	{
+		// Is there a cached result?
+		if (isset($this->data_cache['table_indexes'][$table]))
+		{
+			return $this->data_cache['table_indexes'][$table];
+		}
+	
+		if (FALSE === ($sql = $this->_list_indexes($table)))
```
Comment (tianhe1986) [inline system/database/drivers/mysqli/mysqli_driver.php]: How could it return FALSE since your code is "return 'SHOW INDEXES FROM '.$this->protect_identifiers($table, TRUE, NULL, FALSE)" in function _list_indexes?

File: system/database/drivers/mysqli/mysqli_driver.php [hunk 2]
```diff
@@ -484,6 +499,91 @@ public function field_data($table)
...
+			{
+				if (isset($row['Key_name']))
+				{
+					$key = 'Key_name';
+				}
+				else
+				{
+					// We have no other choice but to just get the third element's key.
+					$temp_array = array_keys($row);
+					$key = $temp[2];
```
Comment (tianhe1986) [inline system/database/drivers/mysqli/mysqli_driver.php]: Maybe "$key = $temp_array[2];" ?

Comment (pkarunyu) [inline system/database/drivers/mysqli/mysqli_driver.php]: @tianhe1986, yes, `$key = $temp_array[2];` is correct! Thanks for that.

File: system/database/drivers/mysqli/mysqli_driver.php (hunk 1 above)
Comment (pkarunyu) [inline system/database/drivers/mysqli/mysqli_driver.php]: @tianhe1986, I presume you are referring to the line `if (FALSE === ($sql = $this->_list_indexes($table)))`?

If yes, you will note that I tried to follow the convention used for the function `list_fields` in [DB_Driver.php](https://github.com/bcit-ci/CodeIgniter/blob/develop/system/database/DB_driver.php#L1246)

Comment (tianhe1986) [inline system/database/drivers/mysqli/mysqli_driver.php]: @pkarunyu 
Yes, you have followed the convention used for the function list_fields in DB_Driver.php, 
but there is still a difference.

In DB_Driver.php, `_list_columns` is undefined(maybe it is also a bug, it should be defined and return FALSE),  and the return value depends on the overridden function in subclass, which may return FALSE.

In your code,  you called `_list_indexes` defined in the same class, which does not return FALSE and would rarely be overridden.

In theory, you are right, but I suggest that you define  `_list_indexes` in DB_Driver.php and override it in each subclass.

Comment (pkarunyu) [inline system/database/drivers/mysqli/mysqli_driver.php]: @tianhe1986, `list_fields` is defined in [DB_driver.php](https://github.com/bcit-ci/CodeIgniter/blob/develop/system/database/DB_driver.php#L1246), but it calls `_list_columns` defined in the individual sub-classes, e.g. [mysqli_driver.php](https://github.com/bcit-ci/CodeIgniter/blob/develop/system/database/drivers/mysqli/mysqli_driver.php#L445). 

However, if you look at the function `list_fields` in DB_driver.php, there is this bit...

```php
// Do we know from where to get the column's name?
if ( ! isset($key))
{
	if (isset($row['column_name']))
	{
		$key = 'column_name';
	}
	elseif (isset($row['COLUMN_NAME']))
	{
		$key = 'COLUMN_NAME';
	}
	else
	{
		// We have no other choice but to just get the first element's key.
		$key = key($row);
	}
}
```

Extrapolating that logic to `list_indexes` would mean that we would have a new function in DB_driver called `list_indexes`, which would call a respective `_list_indexes` in the individual sub-classes. And herein is the dilemma.

In MySQL 5.5 which I tested on, the query `SHOW INDEXES FROM sakila.actor` gives the following:

mysql sakila actor show indexes

Note that the column with the definitive name of the index is the third column, named `Key_name`.

For PostgreSQL 9.5 which I also tested on, the query `SELECT * FROM pg_indexes WHERE tablename = 'actor'` gives the following:

postgresql sakila actor show indexes

Note that the column with the definitive name of the index is the third column, named `indexname`.

Based on these two drivers, then the code block I mentioned above could be re-written thus:

```php
// Do we know from where to get the column's name?
if ( ! isset($key))
{
	if (isset($row['Key_name']))
	{
		$key = 'Key_name';
	}
	elseif (isset($row['indexname']))
	{
		$key = 'indexname';
	}
	else
	{
		// We have no other choice but to just get the third element's key.
		$temp_array = array_keys($row);
		$key = $temp_array[2];
		unset($temp_array);
	}
}
```

In other words, we would have an `elseif` for each database driver, and thus far there are 12.

Comment (tianhe1986) [inline system/database/drivers/mysqli/mysqli_driver.php]: @pkarunyu 

I understand it.

How about keeping `list_indexes` in DB_driver.php with calling a new function defined in each individual sub-classes to get the name of the index ?

Comment (pkarunyu) [inline system/database/drivers/mysqli/mysqli_driver.php]: @tianhe1986, 
> How about keeping list_indexes in DB_driver.php with calling a new function defined in each individual sub-classes to get the name of the index?

That would probably work and seems to be the more elegant way of implementing it.
